server/src/main/java/org/opensearch/search/aggregations/AggregationCollectorManager.java

```diff
@@ -66,7 +66,7 @@ public ReduceableSearchResult reduce(Collection<Collector> collectors) throws IO
         // Reduce the aggregations across slices before sending to the coordinator. We will perform shard level reduce iff multiple slices
         // were created to execute this request and it used concurrent segment search path
         // TODO: Add the check for flag that the request was executed using concurrent search
-        if (collectors.size() > 1) {
+        if (collectors.size() >= 1) {
```

Comment:
This change does not seem to be correct: when there is only one collector, `InternalAggregations.reduce` is not needed.

Reply:
The problem that my change is addressing is `shard_size` and `shard_min_doc_count` were being evaluated at the slice level instead of at the shard level during concurrent search. Specifically, this was done during `buildAggregation`. The solution I came up with in this PR is to move the evaluation from `buildAggregation` to `reduce`, so when we do the shard level reduce we apply `shard_size` and `shard_min_doc_count` there. This line specifically is needed because if there is only 1 segment slice and we do not call `reduce`, then we will not apply `shard_size` and `shard_min_doc_count` in that case. I will add an explanation for that to the code comment though!

Reply:
In this case we don't need the `else` condition at all.

```
if (collectors.isEmpty() == true) {
    return <empty result>;
}
```

Reply:
Actually now I recall what the TODO: comment above meant. The reduce will be called with single collector for non concurrent case as well. We don't want to perform reduce on internal aggregations for non concurrent case (to keep the behavior same as today) and also potentially optimizing for concurrent single slice case. We need to swap this check with if concurrent vs non concurrent path is used and given we have now a use-case to perform reduce with single slice

Reply:
> We don't want to perform reduce on internal aggregations for non concurrent case

I'm wondering now if we should just change this. If we *do* call reduce for the non-concurrent case, then we could skip doing any size/docCount thresholds in `buildAggregations` and do it all in `reduce`. The obvious drawback is that it's risky to change both the concurrent and non-concurrent behavior at the same time.